### PR TITLE
Move cbmc_solvers as solver_factory to new goto-checker module [blocks: 3564, 3565]

### PR DESCRIPTION
**Pull request by peterschrammel**
First of a series of PRs relating to #2883

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [n/a] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [n/a] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 57c66e2332fa6c38bbf78b8ee4aab49da9a6da06).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94355641
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by peterschrammel** [comment]
TG bump is passing

File: src/goto-checker/solver_factory.cpp [hunk 1]
```diff
@@ -118,55 +116,55 @@ std::unique_ptr<cbmc_solverst::solvert> cbmc_solverst::get_bv_refinement()
...
   info.output_xml = output_xml_in_refinement;
 
   // we allow setting some parameters
   if(options.get_bool_option("max-node-refinement"))
-    info.max_node_refinement=
+    info.max_node_refinement =
       options.get_unsigned_int_option("max-node-refinement");
 
-  info.refine_arrays=options.get_bool_option("refine-arrays");
-  info.refine_arithmetic=options.get_bool_option("refine-arithmetic");
```
Comment (romainbrenguier) [inline src/goto-checker/solver_factory.cpp:131]: some formatting changes unrelated to the commit are mixed here

**Review by romainbrenguier — CHANGES_REQUESTED** [review]

File: src/goto-checker/solver_factory.h [hunk 2]
```diff
@@ -50,19 +50,18 @@ class cbmc_solverst
     {
     }
 
-    explicit solvert(std::unique_ptr<prop_convt> p):prop_conv_ptr(std::move(p))
+    explicit solvert(std::unique_ptr<prop_convt> p)
```
Comment (romainbrenguier) [inline src/goto-checker/solver_factory.h:53]: Not directly related to this PR but I find it a bit strange that the product of the factory is defined in the factory itself. And it seems also this should have been factory functions rather than a factory class as most fields are constant and I guess it's only used to produce one solver.

Comment (peterschrammel) [reply src/goto-checker/solver_factory.h:53]: Noted here: https://github.com/diffblue/cbmc/issues/3574

File: src/goto-checker/solver_factory.cpp (hunk 1 above)
Comment (peterschrammel) [reply src/goto-checker/solver_factory.cpp:131]: Now the split should be correct. Please re-check.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 5af9fedcfa77cf95732dfe8f8697c8120443c600).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94870273
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: src/cbmc/bmc.cpp [hunk 3]
```diff
@@ -566,12 +567,12 @@ int bmct::do_language_agnostic_bmc(
...
     {
-      cbmc_solverst solvers(
+      solver_factoryt solvers(
         opts,
         symbol_table,
         ui,
         ui.get_ui() == ui_message_handlert::uit::XML_UI);
-      std::unique_ptr<cbmc_solverst::solvert> cbmc_solver;
+      std::unique_ptr<solver_factoryt::solvert> cbmc_solver;
       cbmc_solver = solvers.get_solver();
```
Comment (romainbrenguier) [inline src/cbmc/bmc.cpp]: :pick: assignment could be on same line as declaration

**Review by romainbrenguier — APPROVED** [review]

File: src/cbmc/bmc.cpp (hunk 3 above)
Comment (peterschrammel) [reply src/cbmc/bmc.cpp]: Done

**Review by allredj — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 699fcc2cc17b6b77dfdf5453941da2ce8134d759).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94956330
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: src/cbmc/bmc.cpp [hunk 4]
```diff
@@ -519,12 +520,12 @@ int bmct::do_language_agnostic_bmc(
...
   {
     {
-      cbmc_solverst solvers(
+      solver_factoryt solvers(
         opts,
         symbol_table,
         ui,
         ui.get_ui() == ui_message_handlert::uit::XML_UI);
-      std::unique_ptr<cbmc_solverst::solvert> cbmc_solver;
+      std::unique_ptr<solver_factoryt::solvert> cbmc_solver;
```
Comment (tautschnig) [inline src/cbmc/bmc.cpp]: As this is also done below: assign directly in the declaration.

File: src/goto-checker/solver_factory.h [hunk 5]
```diff
@@ -1,16 +1,16 @@
...
 \*******************************************************************/
 
 /// \file
-/// Bounded Model Checking for ANSI-C + HDL
+/// Solver Factory
 
-#ifndef CPROVER_CBMC_CBMC_SOLVERS_H
-#define CPROVER_CBMC_CBMC_SOLVERS_H
+#ifndef CPROVER_GOTO_CHECKER_SOLVER_FACTORY_H
+#define CPROVER_GOTO_CHECKER_SOLVER_FACTORY_H
```
Comment (tautschnig) [inline src/goto-checker/solver_factory.h:13]: Wishlist, possibly to be done in a separate PR: could a lot of includes be purged from the header file and the implementation of several methods be moved to the .cpp file? Really nothing needs to be inlined for performance reasons. With such a move we would eventually end up with a clean and readable _interface_.

**Review by tautschnig — APPROVED** [review]
Looks great to me, just more cleanup-wishlist-requests.

**Review by smowton — APPROVED** [review]

File: src/goto-checker/solver_factory.h (hunk 5 above)
Comment (peterschrammel) [reply src/goto-checker/solver_factory.h:13]: https://github.com/diffblue/cbmc/pull/3588

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 700d7bbd0743739b2948c654be9579451d4a7614).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95012430
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.
